File: 455.分发饼干.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <math.h>

using namespace std;
/*
 * @lc app=leetcode.cn id=455 lang=cpp
 *
 * [455] 分发饼干
 */

// @lc code=start
class Solution
{
public:
    int findContentChildren(vector<int> &g, vector<int> &s)
    {
        // 先对数组进行排序
        sort(g.begin(), g.end());
        sort(s.begin(), s.end());
        int num = 0;

        // 空饼干数组处理
        if (s.size() == 0)
        {
            return num;
        }

        // 最大饼干的位置
        int index = s.size() - 1;

        for (int i = g.size() - 1; i >= 0; i--)
        {
            // 从后往前遍历：找到胃口最大小孩的位置
            // 饼干大小 > 小孩胃口，满足条件
            // 继续遍历下一个小孩，index--，去掉刚才已使用的饼干
            if (index >= 0 && s[index] >= g[i])
            {
                num++;
                index--;
            }
        }

        return num;
    }
};
// @lc code=end

File: 贪心算法/easy/455.分发饼干.ts
/*
 * @lc app=leetcode.cn id=455 lang=typescript
 *
 * [455] 分发饼干
 */

// @lc code=start
function findContentChildren(g: number[], s: number[]): number {
  let gIndex = 0
  // 先进行升序排列
  g.sort((a, b) => a - b)
  s.sort((a, b) => a - b)

  // 局部最优解，小饼干先满足小胃口的人
  // 因为是升序，小饼干如果满足同时可以满足大胃口和小胃口
  // 如果满足了大胃口，那么后续的饼干需要去满足更大胃口的
  // 如果满足了小胃口，那么后续的饼干直系要满足当前的大胃口，优于上面
  for (let i = 0; i < s.length; i++) {
    // 这里没有采用两个for轮循
    // 外部遍历饼干，内部采用自增的方式进行遍历小孩
    // gIndex控制小孩数组的下标
    // 小孩被满足，i++,gIndex++，都走到下一个
    // 小孩没被满足，只i++寻找下一个饼干
    if (gIndex < g.length && g[gIndex] <= s[i]) {
      // 满足条件，当前sIndex小孩被满足，跳到下一个小孩
      gIndex++;
    }
  }

  return gIndex
};
// @lc code=end

